src/net_processing.cpp

```diff
@@ -1678,6 +1699,63 @@ void static ProcessGetBlockData(CNode& pfrom, const CChainParams& chainparams, c
     }
 }
 
+class CMWEBLeafsetMsg
+{
+public:
+    CMWEBLeafsetMsg() = default;
+    CMWEBLeafsetMsg(uint256 block_hash_in, BitSet leafset_in)
+        : block_hash(std::move(block_hash_in)), leafset(std::move(leafset_in)) { }
+
+    SERIALIZE_METHODS(CMWEBLeafsetMsg, obj) { READWRITE(obj.block_hash, obj.leafset); }
+
+    uint256 block_hash;
+    BitSet leafset;
+};
+
+static void ProcessGetMWEBLeafset(CNode& pfrom, const CChainParams& chainparams, const CInv& inv, CConnman& connman)
+{
+    ActivateBestChainIfNeeded(chainparams, inv);
+
+    LOCK(cs_main);
+    CBlockIndex* pindex = LookupBlockIndex(inv.hash);
+    if (!pindex) {
+        return;
+    }
+
+    // TODO: Add an outbound limit
+
+    // Avoid leaking prune-height by never sending blocks below the NODE_NETWORK_LIMITED threshold
+    if (::ChainActive().Tip()->nHeight - pindex->nHeight > MAX_MWEB_LEAFSET_DEPTH) {
+        LogPrint(BCLog::NET, "Ignore block request below MAX_MWEB_LEAFSET_DEPTH threshold from peer=%d\n", pfrom.GetId());
+
+        // disconnect node and prevent it from stalling (would otherwise wait for the MWEB leafset)
+        if (!pfrom.HasPermission(PF_NOBAN)) {
+            pfrom.fDisconnect = true;
+        }
+
+        return;
+    }
+
+    // Pruned nodes may have deleted the block, so check whether
+    // it's available before trying to send.
+    if (pindex->nStatus & BLOCK_HAVE_DATA && pindex->nStatus & BLOCK_HAVE_MWEB) {
```

Comment:
Do we need to disconnect if this fails?

Reply:
Yea, probably should. Fixed.

Reply:
ProcessGetBlockData just ignores in this case, and that's what I modeled this function after. I really don't know why we don't always disconnect when something goes wrong. There are so many inconsistencies in how net_processing.cpp handles errors :). Fortunately, the blocks shouldn't ever be pruned since we only allow these requests for the most recent 10 blocks, but  I went ahead and just added the disconnect for completion.